### ISSUE TITLE
serial port open doesnt connect to browser after building app

### ISSUE THREAD

**Issue by d00m303** [xclud/dart_serial#3]
Hello.
I have problem with getting connection to browser. When i run debug version with flutter on localhost i can connect and send/recive data.
But after building same app i cant connect to browser. Device request is opening, im clicking open port and it doesnt claim interface as it does with debug version on localhost.
Am i missing something? Many thank for help ;)
Greets!

**Comment by xclud** [comment]
Are you browsing the web page with `https` scheme? https is required.

> Secure context: This feature is available only in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (HTTPS), in some or all [supporting browsers](https://developer.mozilla.org/en-US/docs/Web/API/SerialPort#browser_compatibility).

https://developer.mozilla.org/en-US/docs/Web/API/SerialPort

**Comment by d00m303** [comment]
Hi. Yes i have working certificate, with https.
Tried on two servers with cert. 
Still i can open port (i see paired device) but it doesnt connect (there is icon on tab when its actually connected but in my case its only in debug mode locally). But thanks for reply i will try with other hostings maybe. Greets!;)

**Comment by d00m303** [comment]
Just checked on my friends commercial servers also and still cant connect.
Im using example app, it runs on external server. When i click open i can choose device but when i click connect nothing happens, there is no typical icon on chrome tab that device is connected.
It works local when i debug app in flutter, starts chrome and all works fine, but not on external server.
Prolly im doing something wrong so u can delete post
Many thanks for help! ;)

**Comment by xclud** [comment]
What version of flutter are you using?

**Comment by d00m303** [comment]
Flutter 2.13.0-0.4.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 25caf1461b (2 tygodnie temu) • 2022-05-05 14:23:09 -0700
Engine • revision c5caf749fe
Tools • Dart 2.17.0 (build 2.17.0-266.8.beta) • DevTools 2.12.2

**Comment by xclud** [comment]
I will investigate this. If you attach some code it will help a lot.

**Comment by d00m303** [comment]
Im using Your example from pub.dev ;)

**Comment by xclud** [comment]
https://github.com/flutter/flutter/issues/104266

**Comment by AkramYns** [comment]
i have the same porbleme even if i host the web app in xampp server

**Comment by cw-20021366** [comment]
I have the same problem.

I checked the log with try&catch for _openPort in your example.
///////////////////
  Future<void> _openPort() async {
    try {
      final port = await window.navigator.serial.requestPort();
      await port.open(SerialOptions(baudRate: 9600));
      _port = port;
    } catch (e) {
      debugPrint('$e');
    }
  }
//////////////////////

then, I got this log
//////////////////////
TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined.
//////////////////////

Could this have something to do with "https"?

**Comment by cw-20021366** [comment]
> I have the same problem.
> 
> I checked the log with try&catch for _openPort in your example. /////////////////// Future _openPort() async { try { final port = await window.navigator.serial.requestPort(); await port.open(SerialOptions(baudRate: 9600)); _port = port; } catch (e) { debugPrint('$e'); } } //////////////////////
> 
> then, I got this log ////////////////////// TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined. //////////////////////
> 
> Could this have something to do with "https"?



> I have the same problem.
> 
> I checked the log with try&catch for _openPort in your example. /////////////////// Future _openPort() async { try { final port = await window.navigator.serial.requestPort(); await port.open(SerialOptions(baudRate: 9600)); _port = port; } catch (e) { debugPrint('$e'); } } //////////////////////
> 
> then, I got this log ////////////////////// TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined. //////////////////////
> 
> Could this have something to do with "https"?

It worked by directly modifying the 'main.dart.js' file.
for example...
org :   A.T(A.dd(n.open(A.aGX(115200)),m),$async$ny)
mod : A.T(A.dd(n.open({ baudRate: 115200 }),m),$async$ny)

Could this be a js conversion issue in "flutter"?

**Comment by cw-20021366** [comment]
Modify SerialOptions class in web.dart as below

> > I have the same problem.
> > I checked the log with try&catch for _openPort in your example. /////////////////// Future _openPort() async { try { final port = await window.navigator.serial.requestPort(); await port.open(SerialOptions(baudRate: 9600)); _port = port; } catch (e) { debugPrint('$e'); } } //////////////////////
> > then, I got this log ////////////////////// TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined. //////////////////////
> > Could this have something to do with "https"?
> 
> > I have the same problem.
> > I checked the log with try&catch for _openPort in your example. /////////////////// Future _openPort() async { try { final port = await window.navigator.serial.requestPort(); await port.open(SerialOptions(baudRate: 9600)); _port = port; } catch (e) { debugPrint('$e'); } } //////////////////////
> > then, I got this log ////////////////////// TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined. //////////////////////
> > Could this have something to do with "https"?
> 
> It worked by directly modifying the 'main.dart.js' file. for example... org : A.T(A.dd(n.open(A.aGX(115200)),m),$async$ny) mod : A.T(A.dd(n.open({ baudRate: 115200 }),m),$async$ny)
> 
> Could this be a js conversion issue in "flutter"?

Modify SerialOptions class in web.dart as below:
`@JS()
@anonymous
abstract class SerialOptions {
  /// A positive, non-zero value indicating the baud rate at which serial communication should be established.
  external int get baudRate;
  external set baudRate(int v);

  /// The number of data bits per frame. Either 7 or 8.
  external int get dataBits;
  external set dataBits(int v);

  /// The number of stop bits at the end of a frame. Either 1 or 2.
  external int get stopBits;
  external set stopBits(int v);

  /// The parity mode.
  external String get parity;
  external set parity(String v);

  /// A positive, non-zero value indicating the size of the read and write buffers that should be created.
  external int get bufferSize;
  external set bufferSize(int v);

  external String get flowControl;
  external set flowControl(String v);

  external factory SerialOptions({
    required baudRate,
    dataBits = 8,
    stopBits = 1,
    parity = 'none',
    bufferSize = 255,
    flowControl = 'none',
  });
}`


https://github.com/google/chartjs.dart/blob/master/lib/chartjs.dart

**Comment by cw-20021366** [comment]
> > I have the same problem.
> > I checked the log with try&catch for _openPort in your example. /////////////////// Future _openPort() async { try { final port = await window.navigator.serial.requestPort(); await port.open(SerialOptions(baudRate: 9600)); _port = port; } catch (e) { debugPrint('$e'); } } //////////////////////
> > then, I got this log ////////////////////// TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined. //////////////////////
> > Could this have something to do with "https"?
> 
> > I have the same problem.
> > I checked the log with try&catch for _openPort in your example. /////////////////// Future _openPort() async { try { final port = await window.navigator.serial.requestPort(); await port.open(SerialOptions(baudRate: 9600)); _port = port; } catch (e) { debugPrint('$e'); } } //////////////////////
> > then, I got this log ////////////////////// TypeError: Failed to execute 'open' on 'SerialPort': Failed to read the 'baudRate' property from 'SerialOptions': Required member is undefined. //////////////////////
> > Could this have something to do with "https"?
> 
> It worked by directly modifying the 'main.dart.js' file. for example... org : A.T(A.dd(n.open(A.aGX(115200)),m),$async$ny) mod : A.T(A.dd(n.open({ baudRate: 115200 }),m),$async$ny)
> 
> Could this be a js conversion issue in "flutter"?

Modify SerialOptions class in web.dart as below:

```dart
@JS()
@anonymous
abstract class SerialOptions {
  external factory SerialOptions({
    required int baudRate,
    int dataBits = 8,
    int stopBits = 1,
    String parity = 'none',
    int bufferSize = 255,
    String flowControl = 'none',
  });

  /// A positive, non-zero value indicating the baud rate at which serial communication should be established.
  external int get baudRate;
  external set baudRate(int v);

  /// The number of data bits per frame. Either 7 or 8.
  external int get dataBits;
  external set dataBits(int v);

  /// The number of stop bits at the end of a frame. Either 1 or 2.
  external int get stopBits;
  external set stopBits(int v);

  /// The parity mode.
  external String get parity;
  external set parity(String v);

  /// A positive, non-zero value indicating the size of the read and write buffers that should be created.
  external int get bufferSize;
  external set bufferSize(int v);

  external String get flowControl;
  external set flowControl(String v);
}
```

Reference https://github.com/google/chartjs.dart/blob/master/lib/chartjs.dart

**Comment by xclud** [comment]
This is fixed in v0.0.3. @cw-20021366 Thank you.

I close this issue but please feel free to open a new one.
